### DISCUSSION TITLE
Gradient Descent Using Sinkhorn

### DISCUSSION THREAD

**Discussion by akul-goyal** [PythonOT/POT#416] [Q&A]
Hi, 

I want to implement the following POT example [here](https://pythonot.github.io/auto_examples/backends/plot_unmix_optim_torch.html#sphx-glr-auto-examples-backends-plot-unmix-optim-torch-py), but instead of using ot.emd2 I want to use Sinkhorn. I tried to use ot.sinkhorn2, but that strips away the gradients from the results. Is there a different function I can use that will keep the gradients?

**Comment by rflamary** [comment]
This is a bug we fixed in the current master branch and is not yet released.

I is very stable  and you can install it with:
```pip install -U https://github.com/PythonOT/POT/archive/master.zip``

**Reply by akul-goyal** [reply]
Hi, I just wanted to follow up on this. I updated the library to the current version, but the gradient was still stripped by Sinkhorn. Interestingly, if I converted b and M2 to Pytorch Variables with requires_grad =True, then I could retain the gradient, but w.grad returned None. All the variables (a,b,M2, loss) return none when accessing .grad. Is there anything I can do to get Sinkhorn to work?

**Reply by rflamary** [reply]
we have a release in the work (once some major PR are merged) but at the moment sou should use the git version be cause the current release do not have the fix yet

**Reply by akul-goyal** [reply]
I updated POT using `pip install -U https://github.com/PythonOT/POT/archive/master.zip ` and installed version 0.8.3.dev0. But I am still getting errors.

**Reply by rflamary** [reply]
I'm sorry but on master (the same version), when editing the loss in the example file as such:
```python
def get_loss(w):
    a = torch.mv(H2, w)  # distribution reweighting
    #return ot.emd2(a, b, M2)  # squared Wasserstein 2
    return ot.sinkhorn2(a, b, M2, 0.1)  # squared Wasserstein 2

```

It just works...
![image](https://user-images.githubusercontent.com/1386873/218753030-7c3e53c6-5938-4486-b39a-d792aaf56c1d.png)
![image](https://user-images.githubusercontent.com/1386873/218753157-cdbb63ee-1fae-4d1c-a70c-d05b5803706b.png)

**Reply by akul-goyal** [reply]
Interestingly, when I pass in `ot.sinkhorn2(a, b, M2, 0.1)` it works, but the following parameters strip the gradients `ot.sinkhorn2(a, b, M2, 0.0005, numItermax=10000000)`. I assume a low entropic regularization term causes gradients to disappear?

**Reply by rflamary** [reply]
for such a small value sinkhorn is not numerically stable dnas the gradient can be bad, you should at least use `method='sinkhorn_log'` in order to ensure that the iterations do not diverge. But for such a small epsilon emd2 will be faster on small datasets.
